Remove multi-collection of batches. (#554)
ietf-wg-ppm/draft-ietf-ppm-dap: draft-ietf-ppm-dap.md

* Prio3 ({{Section 7 of !VDAF}}), which allows for aggregate statistics such as
  sum, mean, histograms, etc.

> * Poplar1 ({{Section 8 of !VDAF}}), which allows for finding the most popular

Author: I dropped references to Poplar1 throughout; I'm not sure if there is anything reasonable to replace it with. (Both PINE & Mastic, IIUC, are not yet to the point where we might cite them.)

Reviewer: We could keep the "counting the number of Clients that hold a given string" use case for Poplar1. This is basically a limited form of attributed based metrics (i.e., Prio3Count with attributes).

The time_interval query type is described in {{time-interval-query}}; the
fixed_size query type is described in {{fixed-size-query}}. Future
specifications may introduce new query types as needed (see
{{query-type-reg}}). Implementations are free to implement only a subset of the
available query types.
specifications may introduce new query types as needed (see {{query-type-reg}}).
Implementations are free to implement only a subset of the available query
types.

A query includes parameters used by the Aggregators to select a batch of
reports specific to the given query type. A query is defined as follows:

~~~
opaque BatchID[32];

enum {
  by_batch_id(0),
  current_batch(1),
  (255)
} FixedSizeQueryType;

struct {
  FixedSizeQueryType query_type;
  select (FixedSizeQuery.query_type) {

>     case by_batch_id: BatchID batch_id;

Author: It is OK to remove `by_batch_id` because this fixed-size query type was used only for repeated collection of the same batch with differing aggregation parameters. (Recovering from transient failures in collecting a given batch is enabled via re-querying the same collection job ID.)

Reviewer: SGTM.

size is in part intended to allow room for error, and allow a range of target
batch sizes.

> ### Batch Size Considerations {#batch-size}

Author: I removed this section because:
 * The second & third paragraphs only apply to multiply-collected VDAFs.
 * The first paragraph boils down to "you should probably set `min_batch_size` and `max_batch_size` to the same value, and if you don't the Leader can set batches freely to any size between these two values."

Should we remove the concept of `max_batch_size`?

Reviewer: Let's consider removing `max_batch_size` in a separate issue.

1. To "refine" the input shares into "output shares" that have the desired
   aggregatable form. For some VDAFs, like Prio3, the mapping from input to
   output shares is some fixed, linear operation, but in general the mapping is

>    controlled dynamically by the Collector and can be non-linear.

Reviewer: No I don't think linearity is important. The suggestion looks good to me.

   the same. If so, the Aggregator MUST mark the input share as invalid with
   error `invalid_message`.

1. Check if the report may still be aggregated with the current aggregation
   parameter. This can be done by looking up all aggregation parameters
   previously used for this report and calling

   ~~~

>    Vdaf.is_valid(current_agg_param, previous_agg_params)

Author: Is there a need for `Vdaf.is_valid` anymore? It _might_ be used to validate that an aggregation parameter itself is valid (irrespective of other aggregation parameters), but I think this might be better done at some point in evaluating the VDAF rather than via a separate method, for simplicity's sake.

Reviewer 2: Yes, I agree. The check for whether a report has been previously aggregated at all should suffice, and VDAF methods like `prep_init` or `prep_next` should safely reject invalid aggregation parameters. It's still useful for `Vdaf.is_valid` to be defined, because DAP-HH would use it, but DAP doesn't need to call it anymore.

Reviewer: I think of validity of the agg param as being delegated to decoding on the wire.

   the same. If so, the Aggregator MUST mark the input share as invalid with
   error `invalid_message`.

1. Check if the report may still be aggregated with the current aggregation
   parameter. This can be done by looking up all aggregation parameters
   previously used for this report and calling

   ~~~
   Vdaf.is_valid(current_agg_param, previous_agg_params)
   ~~~

   If this returns false, the input share MUST be marked as invalid with the
   error `report_replayed`.

    * Implementation note: To detect replay attacks, each Aggregator is required
      to keep track of the set of reports it has processed for a given task.
      Because honest Clients choose the report ID at random, it is sufficient to
      store the set of IDs of processed reports. However, implementations may
      find it helpful to track additional information, like the timestamp, so

>       that the storage used for anti-replay can be sharded efficiently.

Reviewer: This note is still relevant I think.

with an error of type "invalidBatchSize". Leaders SHOULD wait for more reports
to be validated and try the collection job again later.

> Next, the Aggregator checks that the batch has not been queried too many times.

Reviewer 3: Is it possible to effect this repetitive query of the same batch by adopting a rate-limiting scheme that is not tightly bundled with the the underlying DAP. Could we make it loosely coupled?

Reviewer: What would be the goal of rate limiting?

The purpose of querying a batch multiple times is to allow a batch to be aggregated with different aggregation parameters. This enables the tree traversal at the heart of Poplar1's solution to heavy hitters: https://datatracker.ietf.org/meeting/interim-2024-ppm-01/materials/slides-interim-2024-ppm-01-sessa-supporting-heavy-hitters-00

Author: This PR removes the ability to query the same batch multiple times. The purpose of the check this comment was made on is to ensure that we do not collect the same batch multiple times, since doing so may violate the privacy of the reports included in the batch. Thus, we need a solution which permanently blocks the undesirable queries, rather than temporarily rate-limiting them.

As Chris notes, before we removed multi-collection in this PR, each query of the same batch used a distinct aggregation parameter. For the Heavy Hitters application, there was a nontrivial rule to check whether a given aggregation parameter could be used given the aggregation parameters already used. Thus, any given solution to the multiple-queries problem would need to be aware of the aggregation parameters in use.

with an error of type "invalidBatchSize". Leaders SHOULD wait for more reports
to be validated and try the collection job again later.

> Next, the Aggregator checks that the batch has not been queried with multiple
> distinct aggregation parameters. If the batch has been queried with more than
> one distinct aggregation parameter, the Aggregator MUST abort with error of type
> "batchQueriedMultipleTimes".

Reviewer 2: I think this could be stated more clearly in terms of the batch already having been collected.
```suggestion
Next, the Aggregator checks that the batch has not already been aggregated. If it
has, then the Aggregator MUST abort with error of type "batchAlreadyAggregated".
```

Note also the error type changes `batchQueriedMultipleTimes => batchAlreadyAggregated`.

the Aggregator needs to check that the query is associated with a known batch

> ID:

Reviewer: nit: Consider merging the text after the only remaining bullet into this paragraph.

   existence of a report does not imply the occurrence of a sensitive event.
1. Aggregators can deny service by refusing to respond to collection requests or
   aggregate share requests.

> 1. Collection requests may leak information beyond the collection results. For

Author: I'm not sure there is a concrete example to cite here anymore, though I suppose there's nothing stopping a singly-collected VDAF with a nontrivial aggregation parameter from leaking information via the aggregation parameter.

Reviewer 2: I think we should leave this in, even without the explicit reference to Poplar1. Even when restricted to collecting batches once, DAP admits many future, unknown VDAFs, and pointing out that VDAFs may leak information (and not just to the collector!) and may need to be composed with DP is a legitimate security consideration. However I don't like the phrasing "Collection requests may leak information beyond the collection results" because it seems tautological. Let's try:

```suggestion
1. Some VDAFs could leak information to either Aggregator or the Collector
   beyond what the protocol intended to learn. It may be possible to mitigate
   such leakages using differential privacy ({{dp}}). 
```
